text/0000-panic-in-drop.md

```diff
@@ -0,0 +1,319 @@
+    } catch {
+        // If dropping field1 panics the keep trying to drop the remaining fields.
+        try {
+            drop_in_place(&mut (*ptr).field2);
+            drop_in_place(&mut (*ptr).field3);
+        } catch {
+            abort();
+        }
+    }
+
+    // Drop the second field.
+    try {
+        drop_in_place(&mut (*ptr).field2);
+    } catch {
+        // If dropping field2 panics the keep trying to drop the remaining fields.
+        try {
+            drop_in_place(&mut (*ptr).field3);
+        } catch {
+            abort();
+        }
+    }
+
+    // Drop the third field.
+    drop_in_place(&mut (*ptr).field3);
+}
+
+// Codegen with -Z panic-in-drop=abort (new behavior)
+// Also, this function is marked as `nounwind` for LLVM optimizations.
+unsafe fn drop_in_place<T>(ptr: *mut T) {
+    // Call the Drop impl if there is one.
+    try {
+        <T as Drop>::drop(&mut *ptr);
+    } catch {
+        // If the Drop impl panics then abort the process.
+        abort();
+    }
+
+    // Drop the fields: no unwind guards are needed since drop_in_place never
+    // unwinds.
+    drop_in_place(&mut (*ptr).field1);
+    drop_in_place(&mut (*ptr).field2);
+    drop_in_place(&mut (*ptr).field3);
+}
+```
+
+One important restriction is that all crates that are linked together must have the same `panic-in-drop` setting. This is necessary so that the compiler can assume that `drop_in_place<T>` never unwinds, even for `T` types defines in other crates.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+## Deferring execution
+
+A common pattern is to use `Drop` to defer the execution of arbitrary code until the end of a scope. One example of this is the `scopeguard` crate which provides a `defer!` macro. If code within a `defer!` panics then this would abort the process.
+
+Already today, panicking in a `defer!` has the downside of causing a double-panic if the deferred code is invoked as part of an unwind. This is already a problem today: for example, rustc's `delay_span_bug` will defer a panic until the end of a session and only emit it if no other errors occurred. However this can sometimes lead to the compiler double-panicking if rustc is already unwinding due to a previous `span_bug` or `panic`. However if this RFC is accepted then this problem might become more frequent since they will occur in the normal execution path.
+
+## Reporting errors when dropping
+
+Some types can trigger error conditions when dropped. Consider the example of `BufWriter` which needs to flush its buffer when dropped, but where the writing to the underlying `Writer` is a potentially faillible operation. Since it is not possible to "cancel" a drop and return an error, this leaves only 2 possibilities: panic or ignore the error.
+
+The preferred approach is the one used by `BufWriter`: it ignores errors when dropping. If precise error reporting is required then the user can explicitly flush the `BufWriter` and report any errors before dropping the buffer.
```

Comment:
I think it would be worth having some discussion here about what users can do to find current locations where this happens, if they care - do we have clippy lints or something for this? Do we think it's a good idea for such lints (perhaps allow by default) to be added, with an attribute on types/drop impls to signal them as significant?

It feels pretty strange to me to distinguish the flushing on drop from regular write calls in BufWriter (just following that example...); this seems like another case of subtle bugs, if not UB. It's not obvious to me that ignoring the error is the right thing to do, in comparison to panicking (and aborting), which at least let's the user know why they're silently losing data.

In general, I think a clearer answer here would be a good idea - I'm not very comfortable with "just ignore errors silently" as standard recommended practice. One option could be a way to trigger the panic hook (or some similar hook) without actually invoking the panic machinery, which gives opportunity to log the error at least.

Do we have evidence for what other languages do here? I would in particular be interested in GC'd languages which have in some sense "even worse" problems here in the sense of not even being able to reasonably panic (since the stack can be long gone by the time the GC runs)...

Reply:
> Do we have evidence for what other languages do here? I would in particular be interested in GC'd languages which have in some sense "even worse" problems here in the sense of not even being able to reasonably panic (since the stack can be long gone by the time the GC runs)...

- [JDK 18 `Object#finalize`](https://docs.oracle.com/en/java/javase/18/docs/api/java.base/java/lang/Object.html#finalize()): 
  > **Deprecated, for removal: This API element is subject to removal in a future version.** _Finalization is deprecated and subject to removal in a future release._ [...] If an uncaught exception is thrown by the finalize method, the exception is ignored and finalization of that object terminates. [...] Any exception thrown by the `finalize` method causes the finalization of this object to be halted, but is otherwise ignored. [...] Finalizer invocations are not automatically chained, unlike constructors. If a subclass overrides `finalize` it must invoke the superclass finalizer explicitly.
  
  Unwinds from finalization are silently discarded. No finalization of an object is done other than the single `Object#finalize` call. See also [JEP 421](https://openjdk.org/jeps/421) which deprecated finalization for removal and discusses the issues / alternatives.

- [Python 3 `object.__del__(self)`](https://docs.python.org/3/reference/datamodel.html#object.__del__):
  > If a base class has a [__del__()](https://docs.python.org/3/reference/datamodel.html#object.__del__) method, the derived class’s [__del__()](https://docs.python.org/3/reference/datamodel.html#object.__del__) method, if any, must explicitly call it to ensure proper deletion of the base class part of the instance. [...] Due to the precarious circumstances under which [__del__()](https://docs.python.org/3/reference/datamodel.html#object.__del__) methods are invoked, exceptions that occur during their execution are ignored, and a warning is printed to `sys.stderr` instead.
  
  Unwinds from finalization are discarded after printing a warning message to stderr. No finalization of an object is done other than the single `object.__del__(self)` call.
  
- [Swift `deinit`](https://docs.swift.org/swift-book/LanguageGuide/Deinitialization.html):
  > Superclass deinitializers are inherited by their subclasses, and the superclass deinitializer is called automatically at the end of a subclass deinitializer implementation.
  
  Note that Swift is not a tracing GC; `deinit` is deterministically called when the [all strong references are released](https://docs.swift.org/swift-book/LanguageGuide/AutomaticReferenceCounting.html#ID50). Via testing, it looks like it is *impossible* to declare `deinit throws`, and any attempt to `throw` will thus be an error.

- [ECMAScript `FinalizationRegistry`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/FinalizationRegistry) is different enough that it's not really useful prior art, and I can't find any documentation of the behavior of throwing in a callback. My best guess is that this would behave the same as an exception from any other global runtime callback; i.e. log the exception and continue to the next event.

That covers the languages I'm familiar enough with to find the relevant documentation.

Reply:
The consensus of GCd languages seems to be to just discard exceptions from finalizers and continue like nothing happened.

Reply:
Note that GC'd languages are basically universally memory-safe. Swift is probably the most informative case, but there is greater subtlety. In Swift, "throwing" is the equivalent of returning `Err` in Rust (it doesn't even use traditional unwinding, although it does have an optimized ABI). 

*Assertions* in swift do not throw, they trap. That is, all assertions are std::process::abort.

So Swift is essentially designed with the functionality of this proposal: you must either completely swallow errors in dtors, or ~immediately abort.